### PR TITLE
add kernel-based jitter source

### PR DESCRIPTION
**Pull request by thillux**
In our measurements with AARCH64-based NICs (Nvidia Bluefield 2), the userspace jitter RNG provided much less entropy than the kernel-based jitter RNG. The reason for this seems to be the high-resolution timestamp cntvct_el0, which only has a resolution of 100 MHz, while the CPU of the NIC clocks at 2.5 GHz. The non-virtualized timestamp is not accesible from user-space. The high resolution PMU timestamp is also not directly accesible from user-space.

As we did not want to spare CPU cycles for the internal timer-thread in jitterentropy, our current proposal is to add this optional kernel-based jitterentropy source.
This source has access to a high-resolution timestamp and provides a higher entropy count in our test samples. (Most likely it uses cntpct_el0).

### CODE REVIEW

**Comment by thillux** [comment]
After I had a look at the ARM ELs again, cntpct_el0 is may also directly available from user-space, if the CPU is configured accordingly.

**Comment by ORichterSec** [comment]
We have made two observations when looking at the entropy of `jitterentropy`.
1. We have evaluated the entropy of the jitterentropy source with the tool jitterentropy.hashtime (under tests/raw-entropy/recording_userspace https://github.com/smuellerDD/jitterentropy-library/blob/master/tests/raw-entropy/recording_userspace/jitterentropy-hashtime.c).
2. We sampled the raw entropy of jitterentropy_rng via patching the kernel with following Patch:
```
diff --git a/crypto/jitterentropy.c b/crypto/jitterentropy.c
index 227cedfa4f0a..56dc4f593f9a 100644
--- a/crypto/jitterentropy.c
+++ b/crypto/jitterentropy.c
@@ -520,7 +520,10 @@ static int jent_measure_jitter(struct rand_data *ec)
 	stuck = jent_stuck(ec, current_delta);
 
 	/* Now call the next noise sources which also injects the data */
-	jent_lfsr_time(ec, current_delta, 0, stuck);
+	//jent_lfsr_time(ec, current_delta, 0, stuck);
+
+	/* copy entropy directly out for testing */
+	ec->data = current_delta;
 
 	return stuck;
 }
```
and then called the kcapi-rng binary as follows:
kcapi-rng -n jitterentropy_rng -b 8 --hex >> kc.out

On both output files we used the 8 least (jitterentropy.hashtime) / most (kcapi) significant bits to calculate the resulting entropy with the tool djent (https://github.com/dj-on-github/djent).

The above methodology yielded following results:
For a small sample size (5000 samples) the result was as follows:
jitterentropy: 0.324 bit min-entropy/bit (_single_ collumn of jitterentropy.hashtime output)
kcapi: 0.809 bit min-entropy/bit

**Comment by smuellerDD** [comment]
Just to clarify: you are saying that the kernel has access to a faster clock source? I.e. the user space code `asm volatile("mrs %0, cntvct_el0" : "=r" (ctr_val));` can only access a slower clock compared to the random_get_entropy() call in the kernel?

If so, then your patch is the only solution.

File: esdm/esdm_es_jent_kernel.c [hunk 1]
```diff
@@ -0,0 +1,125 @@
+/*
+ * ESDM Fast Entropy Source: Linux jitter-based entropy source
+ *
+ * Copyright (C) 2023, Stephan Mueller <smueller@chronox.de>
```
Comment (smuellerDD) [inline esdm/esdm_es_jent_kernel.c:4]: I never asked to hijack your code. :-)

File: esdm/esdm_es_jent_kernel.h [hunk 2]
```diff
@@ -0,0 +1,32 @@
+/*
+ * Copyright (C) 2023, Stephan Mueller <smueller@chronox.de>
```
Comment (smuellerDD) [inline esdm/esdm_es_jent_kernel.h:2]: dto.

File: esdm/esdm_es_mgr_cb.h [hunk 3]
```diff
@@ -55,6 +55,9 @@ enum esdm_external_es {
 #endif
 #ifdef ESDM_ES_HWRAND
 	esdm_ext_es_hwrand,			/* Linux /dev/hwrng */
+#endif
+#ifdef ESDM_ES_HWRAND
```
Comment (smuellerDD) [inline esdm/esdm_es_mgr_cb.h:59]: ifdef ESDM_ES_JENT_KERNEL ?

**Review by smuellerDD — CHANGES_REQUESTED** [review]
I have not commented on the KNF coding style. You may not need to reformat it as I am going to add a clang-format definition shortly anyway. But for future patches, it would be great to apply KNF.

With the small changes, I am happy to apply the code.

**Comment by thillux** [comment]
> Just to clarify: you are saying that the kernel has access to a faster clock source? I.e. the user space code `asm volatile("mrs %0, cntvct_el0" : "=r" (ctr_val));` can only access a slower clock compared to the random_get_entropy() call in the kernel?
> 
> If so, then your patch is the only solution.

It seems, the kernel can access the CP15 timer, which is more fine-grained in comparison to the generic timer registers (cntvct_el0 and cntpct_el0). cntvct_el0 and cntpct_el0 should increment with the same frequency (contrary to what I suggested above). Therefore it's useful, at least for platforms where such kernel/user-space asymmetry exists for timers, to have access to the kernel-based jitter source.

I'll rebase and cleanup this PR tomorrow.

**Comment by thillux** [comment]
I changed the copyright on files mainly authored by me, rebased and applied formatting. Should be ready for your review.

**Comment by smuellerDD** [comment]
One final question before merge: what is the purpose of the change to common/
logger.c which apparently contains a format change which seemingly is not KNF-
compliant?

**Comment by thillux** [comment]
The format change in kernel logger was caused by your formatting script. Feel free to drop this commit.

**Comment by smuellerDD** [comment]
How interesting. This change did not appear for me using my clang-format version - is that tool applying configurations differently? Anyhow, I would like to drop that for now and rerun my script later. If it re-appears, I will apply it.

**Comment by thillux** [comment]
I don't think so, that my version behaves differently on configured values. My experience with clang-format is, that newer versions change formatting on newly introduced settings, not configured explicitely in the projects formatting config,

**Comment by smuellerDD** [comment]
Applied, thanks.

**Comment by smuellerDD** [comment]
Side note: I just re-ran the helper tool and it did not change anything. So, somehow it behaves differently on my system. I have no idea why.

**Pull request closed without merging**